File: client/src/api/scrape.ts
import apiClient from './client'

export type ScrapeTaskStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

export interface ScrapeTask {
  id: number
  platforms: string        // JSON string e.g. '["instagram","youtube"]'
  industry: string
  target_count: number
  status: ScrapeTaskStatus
  progress: number
  found_count: number
  valid_count: number
  error_message: string | null
  created_by: number | null
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
  target_market?: string | null
  search_keywords?: string | null
  competitor_brands?: string | null
}

export interface ScrapeTaskCreate {
  platforms: string[]
  industry: string
  target_count: number
  target_market?: string
  competitor_brands?: string
}

export interface ScrapeInfluencerResult {
  id: number
  nickname: string | null
  email: string
  platform: string | null
  profile_url: string | null
  followers: number | null
  industry: string | null
  bio: string | null
  status: string
  relevance_score?: number | null
  match_reason?: string | null
}

export const scrapeApi = {
  createTask: (data: ScrapeTaskCreate) =>
    apiClient.post<ScrapeTask>('/scrape/tasks', data).then(r => r.data),

  listTasks: () =>
    apiClient.get<ScrapeTask[]>('/scrape/tasks').then(r => r.data),

  getTask: (id: number) =>
    apiClient.get<ScrapeTask>(`/scrape/tasks/${id}`).then(r => r.data),

  getTaskResults: (id: number, sort: 'followers' | 'default' = 'followers') =>
    apiClient
      .get<ScrapeInfluencerResult[]>(`/scrape/tasks/${id}/results`, { params: { sort } })
      .then(r => r.data),
}

export function parsePlatforms(raw: string): string[] {
  try {
    return JSON.parse(raw)
  } catch {
    return [raw]
  }
}

File: server/app/schemas/scrape.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


SUPPORTED_PLATFORMS = {"instagram", "youtube", "tiktok", "twitter", "facebook"}


class ScrapeTaskCreate(BaseModel):
    platforms: List[str] = Field(min_length=1)
    industry: str = Field(min_length=1, max_length=128)
    target_count: int = Field(default=50, ge=1, le=500)
    target_market: str | None = None
    competitor_brands: str | None = None

    def validate_platforms(self) -> "ScrapeTaskCreate":
        for p in self.platforms:
            if p not in SUPPORTED_PLATFORMS:
                raise ValueError(f"Unsupported platform: {p}")
        return self


class ScrapeTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    platforms: str
    industry: str
    target_count: int
    status: str
    progress: int
    found_count: int
    valid_count: int
    error_message: str | None
    target_market: str | None = None
    search_keywords: str | None = None
    competitor_brands: str | None = None
    created_by: int | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ScrapeInfluencerResult(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nickname: Optional[str]
    email: str
    platform: Optional[str]
    profile_url: Optional[str]
    followers: Optional[int]
    industry: Optional[str]
    bio: Optional[str]
    status: str
    relevance_score: float | None = None
    match_reason: str | None = None
